arches/app/views/api/controlled_lists.py

```diff
@@ -0,0 +1,424 @@
+    # Might add slight python overhead? ~12-14 is enough for Getty AAT.
+    prefetch_depth = 14
+
+    terms = []
+    for i in range(prefetch_depth):
+        if i == 0:
+            terms.extend(
+                [
+                    "controlled_list_items",
+                    "controlled_list_items__controlled_list_item_values",
+                    "controlled_list_items__controlled_list_item_images",
+                    "controlled_list_items__controlled_list_item_images__controlled_list_item_image_metadata",
+                ]
+            )
+        elif find_children:
+            terms.extend(
+                [
+                    f"controlled_list_items{'__children' * i}",
+                    f"controlled_list_items{'__children' * i}__controlled_list_item_values",
+                    f"controlled_list_items{'__children' * i}__controlled_list_item_images",
+                    f"controlled_list_items{'__children' * i}__controlled_list_item_images__controlled_list_item_image_metadata",
+                ]
+            )
+    return terms
+
+
+@method_decorator(
+    group_required("RDM Administrator", raise_exception=True), name="dispatch"
+)
+class ControlledListsView(View):
+    def get(self, request):
+        """Returns either a flat representation (?flat=true) or a tree (default)."""
+        lists = (
+            ControlledList.objects.annotate_node_fields(
+                node_ids="pk",
+                node_names="name",
+                nodegroup_ids="nodegroup_id",
+                graph_ids="graph_id",
+                graph_names="graph__name",
+            )
+            .order_by("name")
+            .prefetch_related(*_prefetch_terms(request))
+        )
+
+        flat = str_to_bool(request.GET.get("flat", "false"))
+        permitted = [
+            ng.pk for ng in get_nodegroups_by_perm(request.user, "read_nodegroup")
+        ]
+        serialized = [
+            obj.serialize(flat=flat, permitted_nodegroups=permitted) for obj in lists
+        ]
+
+        return JSONResponse({"controlled_lists": serialized})
+
+
+@method_decorator(
+    group_required("RDM Administrator", raise_exception=True), name="dispatch"
+)
+class ControlledListView(View):
+    def get(self, request, **kwargs):
+        """Returns either a flat representation (?flat=true) or a tree (default)."""
+        list_id = kwargs.get("id")
+        try:
+            lst = ControlledList.objects.prefetch_related(
+                *_prefetch_terms(request)
+            ).get(pk=list_id)
+        except ControlledList.DoesNotExist:
+            return JSONErrorResponse(status=HTTPStatus.NOT_FOUND)
+
+        flat = str_to_bool(request.GET.get("flat", "false"))
+        permitted = [
+            ng.pk for ng in get_nodegroups_by_perm(request.user, "read_nodegroup")
+        ]
+        serialized = lst.serialize(flat=flat, permitted_nodegroups=permitted)
+
+        return JSONResponse(serialized)
+
+    def add_new_list(self, name):
```

Comment:
Could this go anywhere else? Maybe just fold it into `post`?

Reply:
I'm open to different ideas here. This used to be folded into post, guarded under a comment like "# Add new list.", but I figured that extracting a helper with an explicit function name is better than writing a comment and helps keep the POST more understandable. I think you get more of a sense of _You're either inserting or updating_ this way.

So, it's trivial to fold it back into POST, but I'm not certain I see it as an improvement. Unless we want to say that the only methods on these class should be HTTP verbs. But then I'd say just prefix the method name with an underscore to clarify it's a helper.

Now, whether it should be on the model or somewhere else, I have an open mind here too. It's validating user input and returning an HTTP response. The HTTP response belongs here. The validation is probably best in its own layer, but if we're going to truly isolate a new validation layer, I'd argue for using either model forms or DRF. Our use case in the controlled list manager isn't quite complicated enough for DRF just yet, and I was nervous about how I would handle all the custom stuff around recursive children and prefetching and tree depth, some of which aren't model fields.

I'd be open to investigating model forms going forward. Forgetting to call full_clean() when you're not using any other Django validation tools is something that's pretty error-prone. I just don't know that I'd consider it in scope for this current iteration on this PR.

Reply:
I agree that putting this on the model might be a better place for it.  Then the POST method can then call out to the model method and let the POST catch and handle any errors.

Reply:
Sounds like there's consensus to remove these helper methods, so I did that in 359517a69f2a975d3fcdd09d82ae2b90d00a484e.

But I didn't see a way to put them on the model. The new version looks like this:
```py
    def post(self, request, **kwargs):
        data = JSONDeserializer().deserialize(request.body)

        # Update
        if "id" in kwargs:
            raise NotImplementedError

        # Insert
        name = data.get("name", None)
        lst = ControlledList(name=name)
        try:
            lst.full_clean()  # applies default name
        except ValidationError as ve:
            return JSONErrorResponse(
                message="\n".join(ve.messages), status=HTTPStatus.BAD_REQUEST
            )
        lst.save()
        return JSONResponse(lst.serialize(), status=HTTPStatus.CREATED)
```

If I try to extract everything except the error handling into a method, I'd have something like one of these:

**1** -- Create new instance
```py
    def add_new(name):
        return ControlledList(name=name)
```
**2** -- Create new instance and validate
```py
    def add_new(name):
        clist =  ControlledList(name=name)
        clist.full_clean()
        clist.save()
        return clist
```

**1** seems strange to me. I'd prefer to avoid hooks that only make one function call unless we need it to support subclassing
**2** seems wrong to me, because although Model.clean() is a great place to _define_ validation business logic, _performing_ validation logic doesn't belong in the model. It needs to be obvious where the user-input boundaries of your application are, but that becomes more obscured if you mix it into your model.

To extend on what I was saying above, I'd much rather have a separate validation layer like forms or DRF rather than mix it into the models. I was talking to @aarongundel about this the other day and would be happy to continue some conversations around it.